libbeat/cmd/instance/locks/lock.go

```diff
@@ -0,0 +1,259 @@
+			return lock.recoverLockfile()
+		}
+	}
+
+	// Case: the lockfile is locked, but by us. Probably a coding error,
+	// and probably hard to do
+	if pf.Pid == os.Getpid() {
+		// the lockfile was written before the beat started, meaning we restarted and somehow got the same pid
+		// in which case, continue
+		if lock.beatStart.Before(pf.WriteTime) {
+			return fmt.Errorf("lockfile for beat has been locked twice by the same PID, potential bug.")
+		}
+		lock.logger.Debugf("Beat has started with the same PID, continuing")
+		return lock.recoverLockfile()
+	}
+
+	// Check to see if the PID found in the pidfile exists.
+	existsState, err := findMatchingPID(pf.Pid)
+	// Case: we have a lockfile, but the pid from the pidfile no longer exists
+	// this was presumably due to the dirty shutdown.
+	// Try to reset the lockfile and continue.
+	if errors.Is(err, metricproc.ProcNotExist) {
+		lock.logger.Debugf("%s shut down without removing previous lockfile, continuing", lock.beatName)
+		return lock.recoverLockfile()
+	} else if err != nil {
+		return fmt.Errorf("error looking up status for pid %d: %w", pf.Pid, err)
+	} else {
+		// Case: the PID exists, but it's attached to a zombie process
+		// In this case...we should be okay to restart?
+		if existsState == metricproc.Zombie {
+			lock.logger.Debugf("%s shut down without removing previous lockfile and is currently in a zombie state, continuing", lock.beatName)
+			return lock.recoverLockfile()
+		}
+		// Case: we've gotten a lock file for another process that's already running
+		// This is the "base" lockfile case, which is two beats running from the same directory
+		// This is where we'll catch this particular case on Linux, due to Linux's advisory-style locks.
+		return fmt.Errorf("connot start, data directory belongs to process with pid %d", pf.Pid)
+	}
+}
+
+// recoverLockfile attempts to remove the lockfile and continue running
+// This should only be called after we're sure it's safe to ignore a pre-existing lockfile
+// This will reset the internal lockfile handler when it's successful.
+func (lock *Locker) recoverLockfile() error {
+	// File remove may or not work, depending on os-specific details with lockfiles
+	err := os.Remove(lock.fileLock.Path())
+	if err != nil {
+		if runtime.GOOS == "windows" {
+			// retry on windows, the OS can take time to clean up
+			time.Sleep(time.Second)
+			err = os.Remove(lock.fileLock.Path())
+			if err != nil {
+				return fmt.Errorf("tried twice to remove lockfile %s on windows: %w",
+					lock.fileLock.Path(), err)
+			}
+		} else {
+			return fmt.Errorf("lockfile %s cannot be removed: %w", lock.fileLock.Path(), err)
+		}
+
+	}
+	lock.fileLock = flock.New(lock.filePath)
+	return nil
+}
+
+// readExistingPidfile will read the contents of an existing pidfile
+// Will return false,pidfile{},nil if no data is found in the lockfile
+func (lock *Locker) readExistingPidfile() (bool, pidfile, error) {
```

Comment:
Consider dropping the bool and using an error to indicate an empty data. Then callers can check the cause via `errors.Is(err,  io.ErrUnexpectedEOF)` to determine if it's something they should retry. Like

```go
	if len(rawPidfile) == 0 {
		return nil, fmt.Errorf("empty PID file found: %w", io.ErrUnexpectedEOF)
	}
```